nilearn/decomposition/base_msdl.py

```diff
@@ -0,0 +1,511 @@
+    # apply a ridge on dV = Vs - V, thus Y -> this_Y = Y - np.dot(Us_, V)
+
+    solver = partial(linalg.solve, sym_pos=True,
+                     overwrite_a=True, overwrite_b=True)
+
+    if n_atoms < n_samples:
+        # w = inv(X^t X + alpha*Id) * X.T y
+        cov = np.dot(Us_.T, Us_)
+        this_Y = - np.dot(cov, V)
+        this_Y += np.dot(Us_.T, Ys_)
+        cov.flat[::n_atoms + 1] += mu
+        if not column_wise:
+            dVs = solver(cov, this_Y)
+        else:
+            coefs = np.empty((n_atoms, this_Y.shape[1]))
+            for i in range(this_Y.shape[1]):
+                coefs[:, i] = solver(cov, this_Y[:, i])[0]
+            dVs = coefs
+    else:
+        # w = X.T * inv(X X^t + alpha*Id) y
+        this_Y = - np.dot(Us_, V)
+        this_Y += Ys_
+        cov = np.dot(Us_, Us_.T)
+        cov.flat[::n_samples + 1] += mu
+        if not column_wise:
+            dVs = solver(cov, this_Y)
+            dVs = np.dot(Us_.T, dVs)
+        else:
+            coefs = np.empty((n_atoms, this_Y.shape[1]))
+            for i in range(this_Y.shape[1]):
+                coefs[:, i] = solver(cov, this_Y[:, i])[0]
+            dVs = np.dot(Us_.t, coefs)
+    return V + dVs, (dVs ** 2).sum()
+
+
+def _update_V(Vs, prox, alpha, non_penalized=None):
+    V_group = np.mean(Vs, axis=0)
+    V_group[:non_penalized], V_norms = prox(V_group[:non_penalized], alpha)
+    return V_group, V_norms, [((Vs_ - V_group) ** 2).sum() for Vs_ in Vs]
+
+
+from joblib.memory import MemorizedResult
+
+
+def _get(cached_data):
+    if isinstance(cached_data, MemorizedResult):
+        return cached_data.get(), True
+    return cached_data, False
+
+
+def _update_Vs_Us(Ys, Us, V, mu, verbose=0):
+    tYs, is_shelved = _get(Ys)
+
+    # Update Vs
+    Vs_, residuals_dV_ = _update_Vs(tYs, Us, V, mu=mu)
+
+    # Update Us
+    Us_, r2_ = _update_dict(Us, tYs, Vs_, return_r2=True, verbose=verbose)
+
+    # If subject data is shelved, we dispose of it
+    if is_shelved:
+        del tYs
+    return Vs_, residuals_dV_, Us_, r2_
+
+
+def group_dictionary(Ys, n_atoms, prox, alpha, mu, maxit=100, minit=3,
+            tol=1e-8, n_jobs=1, Us_init=None, Vs_init=None, V_init=None,
+            n_shuffle_subjects=None,
+            callback=None, verbose=False, copy=True, non_penalized=None):
+
+    if verbose > 1:
+        print('[MSDL] Entering group dictionary estimation')
+
+    # Avoid integer division problems
+    alpha = float(alpha)
+    mu = float(mu)
+    n_group = len(Ys)
+
+    t0 = 0
+
+    if n_shuffle_subjects is not None:
+        if n_shuffle_subjects * 2 > n_group:
+            raise ValueError('Number of shuffled subjects must be lesser than '
+                             'number of subjects div 2')
+
+    def cost_function():
+        # r2 is the sum of squares residuals
+        return ((sum(r2) + mu * sum(residuals_dV)) / (n_group * 2.)
+                + mu * alpha * np.sum(V_norms))
```

Comment:
At some point, we spoke of replacing
`subject_fit + mu * (group_fit +  alpha * regularization)`
by something like
`subject_fit + mu * ((1 - alpha) group_fit + alpha * regularization)`

Should we do it?